torch/csrc/jit/export.cpp

```diff
@@ -638,6 +642,186 @@ void ScriptModuleSerializer::writeTensorTable(torch::ModelDef* model_def) {
   }
 }
 
+
+struct Pickler {
+  Pickler(ScriptModuleSerializer& serializer) : serializer_(serializer) {}
+
+public:
+  std::vector<char>& stack() {
+    return stack_;
+  }
+  void start() {
+    push(OpCode::PROTO, uint8_t(2));
+
+    // All attributes get pushed into a list and their indices saved in the
+    // module def
+    push(OpCode::EMPTY_LIST);
+    push(OpCode::MARK);
+  }
+  void finish() {
+    push(OpCode::APPENDS);
+    push(OpCode::STOP);
+  }
+
+  uint64_t addIValue(const IValue& ivalue) {
+    if (ivalue.isTensor()) {
+      // Write it to the tensor table
+      push(OpCode::EXT1);
+      push(ExtensionCode::TENSOR);
+      auto tensor_id = serializer_.addTensor(ivalue.toTensor());
+      push(OpCode::BININT, uint32_t(tensor_id));
+    } else if (ivalue.isBlob()) {
+      AT_ERROR("Unsupported IValue type for pickling (Blob)");
+    } else if (ivalue.isTuple()) {
+      pushTuple(ivalue);
+    } else if (ivalue.isDouble()) {
+      pushDouble(ivalue);
+    } else if (ivalue.isFuture()) {
+      AT_ERROR("Unsupported IValue type for pickling (Future)");
+    } else if (ivalue.isInt()) {
+      // TODO: use BININT1/BININT2/LONG if possible/necessary
+      push(OpCode::BININT, uint32_t(ivalue.toInt()));
+    } else if (ivalue.isBool()) {
+      if (ivalue.toBool()) {
+        push(OpCode::NEWTRUE);
+      } else {
+        push(OpCode::NEWFALSE);
+      }
+    } else if (ivalue.isIntList()) {
+      AT_ERROR("Unsupported IValue type for pickling (IntList)");
+    } else if (ivalue.isString()) {
+      pushString(ivalue);
+    } else if (ivalue.isDoubleList()) {
+      AT_ERROR("Unsupported IValue type for pickling (DoubleList)");
+    } else if (ivalue.isBoolList()) {
+      AT_ERROR("Unsupported IValue type for pickling (BoolList)");
+    } else if (ivalue.isTensorList()) {
+      AT_ERROR("Unsupported IValue type for pickling (TensorList)");
+    } else if (ivalue.isGenericList()) {
+      pushList(ivalue);
+    } else if (ivalue.isGenericDict()) {
+      pushDict(ivalue);
+    } else if (ivalue.isObject()) {
+      AT_ERROR("Unsupported IValue type for pickling (Object)");
+    } else if (ivalue.isNone()) {
+      push(OpCode::NONE);
+    } else if (ivalue.isDevice()) {
+      AT_ERROR("Unsupported IValue type for pickling (Device)");
+    } else {
+      AT_ERROR("Unknown IValue type for pickling");
+    }
+
+    return 0;
+  }
+
+private:
+  void pushString(const IValue& ivalue) {
+    auto string = ivalue.toStringRef();
+
+    push(OpCode::BINUNICODE);
+    push(uint32_t(string.size()));
+    stack_.insert(stack_.end(), string.begin(), string.end());
+
+    push(OpCode::BINPUT, memo_id++);
```

Comment:
Needs checks for running out of memo values and using BINPUT vs BINPUT_LONG.

Reply:
Also we are not saving the mapping from the IValue's reference pointer to this memo_id, so we are never correctly issuing BINGETs rather than duplicating the object.